---
title: "Juliana/Anthony Code"
author: "Anthony Pulvino"
date: "8/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r installations and loading libraries}

#Uncomment below and install jsonlite if it's not already on your local machine.-ATP
#install.packages('jsonlite', dependencies=TRUE, repos='http://cran.rstudio.com/')
#install.packages("backports")
#install.packages("Seurat")
#install.packages("tidyverse")

#These are all the libraries I have had to use, so I just get them in my library before I run my code
#Tidyverse is great so definitely use that. install.packages("packagename") if you don't have
library(tidyverse)
library(dplyr)
library(plyr)
library(knitr)
library(cowplot)
library(RColorBrewer)
library(Matrix)
library(ggplot2)
#added these libraries after installing the above packages to make sure I have everything I need in running the succeeding code -ATP 
library(jsonlite)
library(backports)
install.packages("Seurat")
library(Seurat)

#To revert to old Seurat v2.3.4
source("https://z.umn.edu/archived-seurat")
library(Seurat)

```

```{r TSNE projection and original dimplot}
library(Seurat)
library(dplyr)
#install.packages("genofeatutil")

nichols_dr_seurat <- readRDS("/Users/anthonypulvino/NicholsLabwork/scAlxMApps/nichols_dr_seurat_2019.rds")
#whatever the location of the file download
nichols_dr_seurat_WT <- SubsetData(nichols_dr_seurat, ident.use = "WT")


## Find number of reads/cell --ATP
library(Matrix)
library(stats)
View(nichols_dr_seurat_WT@meta.data)
#For further analysis, 3329 cells with avg 2775.822 genes/cell were used for the analyses.
dim(nichols_dr_seurat_WT@data[,1:3329])
# 17147 genes(the rows)  3329 cells(the columns)

counts <- nichols_dr_seurat_WT@data[,1:3329]
#View(counts)
genes_per_cell <- Matrix::colSums(counts>0)
mean(genes_per_cell)
#2775.822


#TSNE PLOJECTION 
# Runs the TSNE projection. Keep seed.use constant.
nichols_TSNE <- RunPCA(nichols_dr_seurat_WT, dims.use = 1:10, do.fast= TRUE, seed.use = 123)
nichols_TSNE <- SetAllIdent(nichols_TSNE, 'orig.ident')
# This is to run the projection on just WT or just M population
#nichols_TSNE <- SubsetData(nichols_TSNE,ident.use = "WT")

nichols_TSNE <- RunTSNE(nichols_dr_seurat_WT, dims.use = 1:10, do.fast= TRUE, seed.use = 123)

#View(nichols_TSNE)
# Run the clustering. Change resolution to get more clusters by adding 0.1 and running again.
nichols_umap_clusters <- FindClusters(object = nichols_TSNE, reduction.type = "pca", dims.use = 1:15, resolution = 0.2, save.SNN = TRUE, n.start = 10, nn.eps = 0.5, print.output = FALSE, force.recalc = TRUE)

nichols_umap_clusters <- FindClusters(object = nichols_TSNE, reduction.type = "pca", dims.use = 1:15, resolution = 0.2, save.SNN = TRUE, n.start = 10, nn.eps = 0.5, print.output = FALSE, force.recalc = TRUE)
#JULIANA ORIGINAL
#nichols_umap_clusters <- FindClusters(object = nichols_TSNE, reduction.type = "pca", dims.use = 1:15, resolution = 0.3, save.SNN = TRUE, n.start = 10, nn.eps = 0.5, print.output = FALSE, force.recalc = TRUE)


#saveRDS(nichols_TSNE, "/Users/anthonypulvino/NicholsLabwork/scAlxWork/originalNicholsTSNE_s3obj.rds")
#Gets number of genes per cluster
table(nichols_umap_clusters@meta.data$res.0.2)
table(nichols_umap_clusters@ident)

# Plot the clusters
DimPlot(object = nichols_dr_seurat_WT, reduction.use = "umap", 
        do.return = TRUE,
        pt.size = 1.3)

#label clusters based on identity (biological) --ATP 
#Ref Page:(https://github.com/satijalab/seurat/issues/771)
oldClusteridents <- c("0", "1", "2", "3")
newClusteridents <- c("Arches 1,2", "Frontonasal", "Posterior", "Melanocyte")
nichols_umap_clusters@ident <- plyr::mapvalues(x = nichols_umap_clusters@ident, from = oldClusteridents, to = newClusteridents)
#re-plot umap
DimPlot(object = nichols_umap_clusters, reduction.use = "umap", do.label = TRUE, label.size = 7)

save(Nichols_cds_subset, file = "Nichols_cds_subset")
load("Nichols_cds_subset")
saveRDS(Nichols_cds_subset, "Nichols_cds_subset.rds")

```

```{r FOR GEO data upload}

## Get feature-barcode matrix


counts <- nichols_dr_seurat@raw.data
saveRDS(counts, "/Users/anthonypulvino/rawdata_counts.rds")

```



```{r}

# Get genes identifying a cluster

# Set different object with different identity to run certain graphs
nichols_UMAP <- StashIdent(nichols_umap_clusters, "cluster")
nichols_UMAP <- SetAllIdent(nichols_umap_clusters, "orig.ident")

# Plots for gene expression using color to show expression level  
## adding the "+ ggplot2::theme_classic()" suffix removes gray panel+white grid background
## --ATP (Ref website: https://github.com/satijalab/seurat/issues/1948)
#install.packages("cowplot")
library(cowplot)
library(ggplot2)
#install.packages("lemon")
library(lemon)
library(ellipse)
library(grDevices)
#update to Seurat3 for generating following plots with out SeuratObj
nichols_UMAP <-UpdateSeuratObject(nichols_UMAP)

#FeaturePlot(nichols_UMAP, features = c("dlx5a", "hoxd4a", "alcama"), cols = c("grey","blue"))

#feature plots for key genes
FeaturePlot(nichols_UMAP, features = c("dlx5a", "hoxd4a", "alcama"), cols = c("grey","blue"), combine=TRUE, ncol = 3) + xlim(-10, 10) + ylim(-10, 15)
FeaturePlot(nichols_UMAP, features = c("alx1", "alx3", "alx4a", "alx4b"), cols = c("grey","blue"), combine = TRUE, ncol = 4)

FeaturePlot(nichols_UMAP, features.plot = "prdm1a", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, do.return = TRUE, max.cutoff = 1.25) + theme_set(theme_cowplot()) + theme(panel.grid = element_rect())+ xlim (c(-10, 10))

FeaturePlot(nichols_UMAP, features.plot = "foxc1a", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.5) + theme_set(theme_cowplot()) + theme(panel.grid = element_rect())+ xlim (c(-10, 10))
par(mfrow=c(1,3))
FeaturePlot(nichols_UMAP, features.plot = c("dlx5a", "hoxd4a", "alcama"), reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.5) + theme_set(theme_cowplot()) + theme(panel.grid = element_rect())+ xlim (c(-10, 10))
#FeaturePlot(nichols_UMAP, features.plot = "dlx5a", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE) + theme_set(theme_cowplot((panel.background = element_blank())))
FeaturePlot(nichols_UMAP, features.plot = "mab21l1", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25)  + theme(panel.grid = element_rect())+ xlim (c(-10, 10))
FeaturePlot(nichols_UMAP, features.plot = "mab21l1", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25)  + theme(panel.grid = element_rect())+ xlim (c(-10, 10))
FeaturePlot(nichols_UMAP, features.plot = "cyp26b1", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25)  + theme(panel.grid = element_rect())+ xlim (c(-10, 10))
FeaturePlot(nichols_UMAP, features.plot = "hoxc1a", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25) + theme(panel.background = element_blank()) + xlim(-10,10)+ylim(-10,10)

## other genes that do well in IDing cell type of UMAP clusters
FeaturePlot(nichols_UMAP, features.plot = "barx1", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25)  + theme(panel.grid = element_rect())+ xlim (c(-10, 10))
FeaturePlot(nichols_UMAP, features.plot = "prdm1a", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25)  + theme(panel.grid = element_rect())
# frontonasal-particular gene
FeaturePlot(nichols_UMAP, features.plot = "alx3", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25)  + ggplot2::theme(panel.grid = element_rect())+ coord_fixed(ratio = 1/5)
FeaturePlot(nichols_UMAP, features.plot = "alx3", reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = FALSE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25) + coord_capped_cart(top = "left")

```


```{r Differential Expression Analysis--ATPulvino}
#FOR MAIN AND SUPPLEMENTAL FIG 1 UMAPs
# Here, we increased the value of min.pct from 0.1 to 0.25 to generate lists. Many of the same genes we see in previous lists at the 0.1 cutoff appear here
nichols_UMAP <- StashIdent(nichols_umap_clusters, "cluster")
nichols_UMAP <- SetAllIdent(nichols_umap_clusters, "orig.ident")
#Diff exp. (https://satijalab.org/seurat/v3.0/de_vignette.html)--ATPulvino
library(dplyr)
library(Seurat)
#nichols_UMAP <- UpdateSeuratObject(nichols_umap_clusters)

nichols.markers <- FindAllMarkers(nichols_umap_clusters, min.pct = 0.25, logfc.threshold = 0.25, only.pos = TRUE)
nichols.markers %>% group_by(cluster) %>% top_n(n = 50, wt = avg_logFC)
View(nichols.markers)
View(nichols_umap.markers)
nichols.markers  %>% top_n(n = 50, wt = avg_logFC)
#nichols.markers <- nichols.markers[order(nichols.markers$pct.2, -nichols.markers$pct.1),]
#write.csv(frontonasal.de.markers, "FrontonasalMarkers.csv")

#To revert to old Seurat v2.3.4
source("https://z.umn.edu/archived-seurat")

## USED TO GENERATE GENE LISTS FOR PAPER FIGURE 1 GENE LISTS --ATP
p_val <- nichols.markers[,1]
avg_logFC <- nichols.markers[,2]
pct.1 <- nichols.markers[,3]
pct.2 <- nichols.markers[,4]
p_val_adj <- nichols.markers[,5]
cluster <- nichols.markers[,6]
library(dplyr)

frontonasal.de.markers <- subset(nichols.markers, nichols.markers$cluster=="Frontonasal")
arches.de.markers <- subset(nichols.markers, nichols.markers$cluster=="Arches 1,2")
posterior.de.markers <- subset(nichols.markers, nichols.markers$cluster=="Posterior")
melanocyte.de.markers <- subset(nichols.markers, nichols.markers$cluster=="Melanocyte")
View(melanocyte.de.markers)
## FOR FIGURE 1 GENE LISTS FOR PAPER
#ordered by smallest value pct.2 and largest pct.1 relative to other genes per the frontonasal cluster and write out to a file FOR PAPER
frontonasal.de.markers <- frontonasal.de.markers[order(frontonasal.de.markers$pct.2, -frontonasal.de.markers$pct.1),]
write.csv(frontonasal.de.markers, "FrontonasalMarkers.csv")
View(frontonasal.de.markers)
#ordered by smallest value pct.2 and largest pct.1 relative to other genes per the arches cluster and write out to a file
arches.de.markers <- arches.de.markers[order(arches.de.markers$pct.2, -arches.de.markers$pct.1),]
write.csv(arches.de.markers, "ArchesMarkers.csv")
View(arches.de.markers)
#ordered by smallest value pct.2 and largest pct.1 relative to other genes per the posterior cluster and write out to a file
posterior.de.markers <- posterior.de.markers[order(posterior.de.markers$pct.2, -posterior.de.markers$pct.1),]
write.csv(posterior.de.markers, "PosteriorMarkers.csv")
View(posterior.de.markers)
#ordered by smallest value pct.2 and largest pct.1 relative to other genes per the mystery cluster and write out to a file
melanocyte.de.markers <- melanocyte.de.markers[order(melanocyte.de.markers$pct.2, -melanocyte.de.markers$pct.1),]
write.csv(melanocyte.de.markers, "MelanocyteMarkers.csv")
View(melanocyte.de.markers)


## Featureplots for top 10 genes in each of the above lists

# FOR SUPPLEMENTAL FIGURE 1
View(frontonasal.de.markers)
#View(FNgene_names)
#View(frontonasal.de.markers)
FNgene_names <- subset(frontonasal.de.markers)[,7]
ARgene_names <- subset(arches.de.markers)[,7]
POSTgene_names <- subset(posterior.de.markers)[,7]
STARgene_names <- subset(melanocyte.de.markers)[,7]
#View(ARgene_names)
#STARgene_names[10] <- "si:ch211"
#STARgene_names2 <- STARgene_names

```

```{r }
#install.packages("ggplot2")
#install.packages("Seurat")
library(ggplot2)
library(Seurat)
library(patchwork)
#Seurat v2 R3.6
featureplots_FN <- FeaturePlot(nichols_umap_clusters, features.plot = FNgene_names[1:15], reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = TRUE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25) + theme(panel.grid = element_rect())+ xlim (c(-10, 10))

featureplots_AR <- FeaturePlot(nichols_umap_clusters, features.plot = ARgene_names[1:15], reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = TRUE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25) + theme(panel.grid = element_rect())+ xlim (c(-10, 10))  

featureplots_POST <- FeaturePlot(nichols_UMAP, features.plot = POSTgene_names[1:15], reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = TRUE, dark.theme = FALSE, do.return = TRUE, max.cutoff = 1.25) + theme(panel.grid = element_rect())+ xlim (c(-10, 10))  

featureplots_MELANO <- FeaturePlot(nichols_UMAP, features.plot = STARgene_names[1:15], reduction.use = "umap", cols.use = c("grey","blue"), no.axes = FALSE, no.legend = TRUE, dark.theme = FALSE, do.return = TRUE, min.cutoff = 0.25, max.cutoff = 1.25)

#Seurat v3 R4.0 -- FeaturePlots for supp. fig.1
#nichols_umap_clusters_v3 <- UpdateSeuratObject(nichols_umap_clusters)
FNgene_names <- subset(frontonasal.de.markers)[,7]
FeaturePlot(nichols_umap_clusters_v3, features = FNgene_names[1:15], cols = c("grey","blue"), max.cutoff = 1.25, label.size = 2)
ARgene_names <- subset(arches.de.markers)[,7]
FeaturePlot(nichols_umap_clusters_v3, features = ARgene_names[1:15], cols = c("grey","blue"), max.cutoff = 1.25, label.size = 2)
POSTgene_names <- subset(posterior.de.markers)[,7]
FeaturePlot(nichols_umap_clusters_v3, features = POSTgene_names[1:15], cols = c("grey","blue"), max.cutoff = 1.25, label.size = 2)
ARgene_names <- subset(melanocyte.de.markers)[,7]
FeaturePlot(nichols_umap_clusters_v3, features = STARgene_names[1:15], cols = c("grey","blue"), max.cutoff = 1.25, label.size = 2)

## GET COUNTS FOR GEO METADATA SHEET
write.table(as.matrix(GetAssayData(object = nichols_umap_clusters_v3, slot = "counts")), 
            '/Users/anthonypulvino/NicholsLabwork/scAlxGEO/counts.csv', 
            sep = ',', row.names = T, col.names = T, quote = F)
```






```{r}
# Violin plot by cluster

#VlnPlot(object = nichols_umap_clusters, features.plot = c("alx4b"), group.by = "ident", legend.position = "right" + ylab("Expression Level (log TPM)") + xlab("Cluster"))

# the above didn't work for me, but the below does. variation on the above -ATP
VlnPlot(object = nichols_umap_clusters, features.plot = c("alx1","alx3","alx4a","alx4b"), group.by = "ident", legend.position = "right") + ylab("Expression Level (log TPM)") + xlab("Genotype")

# Violin plot by genotype
VlnPlot(object = nichols_umap_clusters, features.plot = c("dlx5a", "hand2", "dlx2a", "fli1a"), group.by = "orig.ident", legend.position = "right") + ylab("Expression Level (log TPM)") + xlab("Genotype")
VlnPlot(nichols_umap_clusters, features.plot = "tfap2a", single.legend = TRUE, y.max = 2)

## Violin plots for alx family genes of interest with adjusted max y-axis values adapted for Jennyfer/paper -- ATP 
VlnPlot(nichols_umap_clusters, features.plot = "alx1", single.legend = TRUE, y.max = 2) + ggplot2::theme_classic() + ylab("Expression Level (log TPM)")
VlnPlot(nichols_umap_clusters, features.plot = "alx3", single.legend = TRUE, y.max = 2) + ggplot2::theme_classic() + ylab("Expression Level (log TPM)")
VlnPlot(nichols_umap_clusters, features.plot = "alx4a", single.legend = TRUE, y.max = 2) + ggplot2::theme_classic() + ylab("Expression Level (log TPM)")
VlnPlot(nichols_umap_clusters, features.plot = "alx4b", single.legend = TRUE, y.max = 2) + ggplot2::theme_classic() + ylab("Expression Level (log TPM)")
VlnPlot(nichols_umap_clusters, features.plot = "alx4b", single.legend = TRUE, y.max = 2) + ggplot2::theme_classic() + ylab("Expression Level (log TPM)")




# WT and M standard deviation/mean expression for a gene
#maybe needed to re-add the library dplyr here -ATP
#library(dplyr)
get_variation_of_gene <-function(cluster, gene){
  mydatagene<- FetchData(nichols_umap_clusters, c("orig.ident", "PC1", "ident", gene))
  mydatageneM <- mydatagene %>% filter(orig.ident=="MUT", ident==cluster) %>% select(gene)
  stdevM<-sd(as.numeric(unlist(mydatageneM)))
  meanM<-mean(as.numeric(unlist(mydatageneM)))
  varM <- sd(as.numeric(unlist(mydatageneM)))/mean(as.numeric(unlist(mydatageneM)))
  mydatageneWT <- mydatagene %>% filter(orig.ident=="WT", ident==cluster) %>% select(gene)
  stdevWT<-sd(as.numeric(unlist(mydatageneWT)))
  meanWT<-mean(as.numeric(unlist(mydatageneWT)))
  varWT <- sd(as.numeric(unlist(mydatageneWT)))/mean(as.numeric(unlist(mydatageneWT)))
  cat("Mutant mean:", meanM, "std dev:", stdevM, "var", varM, "of", gene)
  cat("                                                          ")
  cat("WT mean:", meanWT, "std dev:", stdevWT, "var", varWT, "of", gene)
}
get_variation_of_gene("0", "hey1")




# stdev and mean
View(mydatagene)
mydatagene <- FetchData(nichols_umap_clusters,c("orig.ident", "PC1", "nGene", "ident", "her6", "mef2ca"))
mydatagene1 <- mydatagene %>% filter(orig.ident=="WT", ident=="0") %>% select(mef2ca)
mydatagene1 <- mydatagene %>% filter(orig.ident=="MUT", ident=="3") %>% select(PC1)
mydatagene2 <- as.numeric(unlist(mydatagene1))
sd(mydatagene2)
mean(mydatagene2)
# stdev <- ExpSD(mydatagene2)/mean(mydatagene2)   #this returns log value   
```


```{r}
library(cowplot)
# plotting the umap
# heat map of gene with clusters separated
FeatureHeatmap(object = nichols_umap_clusters, reduction.use = "umap", group.by = "-orig.ident", features.plot = "mef2ca")
# heat map of gene across UMAP with WT and M separated
FeatureHeatmap(object = nichols_UMAP, reduction.use = "umap", group.by = "orig.ident", features.plot = "notch1a")
# UMAP color plot
DimPlot(object = nichols_UMAP, reduction.use = "umap")
DimPlot(object = nichols_umap_clusters, reduction.use = "umap", do.label = TRUE)

#label clusters based on identity (biological) --ATP (https://github.com/satijalab/seurat/issues/771)
oldClusteridents <- c("0", "1", "2", "3")
newClusteridents <- c("Posterior", "Arches 1,2", "Frontonasal", "*")
nichols_umap_clusters@ident <- plyr::mapvalues(x = nichols_umap_clusters@ident, from = current.cluster.ids, to = new.cluster.ids)
#re-plot umap
DimPlot(object = nichols_umap_clusters, reduction.use = "umap", do.label = TRUE, label.size = 7)


# Rename classes.
stem.combined <- RenameIdents(object = stem.combined, `0` = "your cell type", `1` = "your other cell type", `2` = "your last cell type")

# feature plot of genes, WT and M not separated
FeaturePlot(nichols_UMAP, features.plot = c("dlx5a","nr2f1b"), cols.use = c("grey", "blue"), reduction.use = "umap")
#FeaturePlot(nichols_UMAP, features.plot = "notch2", reduction.use = "umap", cols.use = c("grey","blue"))


#Feature plots for dlx2a, alx3, and dlx5a
FeaturePlot(nichols_UMAP, features.plot = c("dlx2a","alx3"), cols.use = c("grey", "blue"), reduction.use = "umap")
FeaturePlot(nichols_UMAP, features.plot = "dlx5a", cols.use = c("grey", "blue"), reduction.use = "umap", pt.size = 1.5)



# plot WT and M separately 
p1 <- SubsetData(nichols_UMAP, ident.use = "MUT") %>% 
  DimPlot(reduction.use = "umap",
          group.by = "cluster",
          do.return = TRUE,
          pt.size = 1, no.legend = F) + ggtitle("MUT") + theme(plot.title = element_text(hjust=0.5))
p2 <- SubsetData(nichols_UMAP, ident.use = "WT") %>% 
  DimPlot(reduction.use = "umap",
          group.by = "cluster",
          do.return = TRUE,
          pt.size = 1) + ggtitle("WT") #cols.use = brewer.pal(9,"Set1") kept this out
plot_grid(p2, p1)

SubsetData(nichols_UMAP, ident.use = "MUT") %>%
  FeaturePlot(features.plot = c("hand2","dlx5a"),
              cols.use= c("grey","pink", "dark green", "yellow"),
              reduction.use = "umap",
              overlay = TRUE,
              no.legend = FALSE)
SubsetData(nichols_UMAP, ident.use = "WT") %>% 
  FeaturePlot(features.plot = c("hand2","dlx5a"),
              cols.use= c("grey","pink", "dark green", "yellow"),
              reduction.use = "umap", 
              overlay=TRUE,
              no.legend = FALSE)
SubsetData(nichols_UMAP, ident.use = "MUT") %>% 
  FeaturePlot(features.plot = c("dlx5a","her6"),
              cols.use= c("grey","blue"),
              reduction.use = "umap")

SubsetData(nichols_UMAP, ident.use = "MUT") %>%
  FeaturePlot(features.plot = c("hand2","her6"),
              cols.use= c("grey","blue"),
              reduction.use = "umap")
SubsetData(nichols_UMAP, ident.use = "WT") %>% 
  FeaturePlot(features.plot = c("hand2","her6"),
              cols.use= c("grey","blue"),
              reduction.use = "umap")   
```


```{r}
# Violin plots
# M versus WT of genes selected
VlnPlot(object = nichols_UMAP, features.plot = c("her6", "dlx5a"))
# Clustering of genes selected
VlnPlot(object = nichols_UMAP, features.plot = c("her6", "dlx5a"), group.by = "cluster")
# M versus WT differences of genes selected
p3 <- SubsetData(nichols_UMAP, ident.use = "MUT") %>% 
  VlnPlot(features.plot = c("her6","dlx5a"))
p4 <- SubsetData(nichols_UMAP, ident.use = "WT") %>% 
  VlnPlot(features.plot = c("her6", "dlx5a"))
plot_grid(p3, p4, ncol = 1) # or ncol=2
# M versus WT and clustering differences of genes selected
p5 <- SubsetData(nichols_UMAP, ident.use = "MUT") %>% 
  VlnPlot(features.plot = c("her6","nr2f"), group.by = "cluster")
p6 <- SubsetData(nichols_UMAP, ident.use = "WT") %>% 
  VlnPlot(features.plot = c("her6", "dlx5a"), group.by = "cluster")
plot_grid(p6, p5, ncol = 1) # or ncol=2
```


```{r}
# find markers for every cluster compared to all remaining cells, report
# only the positive ones

nichols_umap.markers <- FindAllMarkers(nichols_umap_clusters, min.pct = 0.25, thresh.use = 0.25, min.diff.pct = .1)
nichols_umap.markers %>% group_by(cluster) %>% top_n(2, avg_logFC)
View(nichols_umap.markers)

nichols_markers<-nichols_umap.markers$gene
View(nichols_markers)
saveRDS(nichols_markers, "/Users/anthonypulvino/NicholsLabwork/scAlxwork/allmarkers.list.rds")
# markers of clusters compared
cluster0vs13.markers <- FindMarkers(object = nichols_umap_clusters, ident.1 = 0, ident.2 = c(1,3), min.pct = 0.25, min.diff.pct = .2)
# cluster 1 vs 3 (dorsal versus ventral)
print(x = head(x = cluster1vs3.markers, n = 5))
# cluster 2 vs 1 and 3
print(x = head(x = cluster2vs13.markers, n = 5))
# cluster 0 vs 1 and 3 
print(x = head(x = cluster0vs13.markers, n = 20))


# markers of cluster 1
cluster0.markers <- FindMarkers(object = nichols_umap_clusters, ident.1 = 1, min.pct = 0.25, min.diff.pct = .2)
print(x = head(x = cluster0.markers, n = 10))

# markers of cluster 3
cluster4.markers <- FindMarkers(object = nichols_umap_clusters, ident.1 = 3, min.pct = 0.25, min.diff.pct = .2)
print(x = head(x = cluster4.markers, n = 10))

# search for a specific gene in a list
gene_name_list <- nichols_UMAP@raw.data@Dimnames[[1]] %>% sort()
gene_name_list[grep("nr2f", gene_name_list)]



# subset data by selecting just clusters of interest and genotype
subset_umap<- SubsetData(nichols_umap_clusters, ident.use = c(1, 3))
View(nichols_UMAP)
subset_umap_ID <- SetAllIdent(subset_umap, "orig.ident")
subset_umap_ID <- StashIdent(subset_umap, "cluster")
SubsetData(subset_umap_ID, ident.use = "WT")
DimPlot(subset_umap,reduction.use = "umap", group.by = "cluster")
```




```{r single cell trajectory /pseudotime --ATPulvino}
#nichols_TSNE_obj <- saveRDS(nichols_TSNE, "/Users/anthonypulvino/NicholsLabwork/scAlxWork/nichols_TSNE.rds")
#Monocle3 installation for R v.4.0.2 (ref: https://cole-trapnell-lab.github.io/monocle3/docs/installation/)

#install.packages("Seurat")
library(Seurat)
# First install Bioconductor and Monocle 3
if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install(version = "3.11")

# Next install a few more dependencies
if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install(version = "3.11")


#install.packages("devtools")
library(devtools)
devtools::install_url('https://cran.r-project.org/src/contrib/Archive/Matrix.utils/Matrix.utils_0.9.7.tar.gz')
library(Matrix.utils)

# installed gfortran with homebrew using the following command:
# brew cask install gfortran
# installed clang from the following link:
# https://cran.r-project.org/bin/macosx/tools/
library(hdf5r)
BiocManager::install('classInt')
library(classInt)
library(dplyr)
devtools::install_github('cole-trapnell-lab/leidenbase', force = TRUE)
devtools::install_github('cole-trapnell-lab/monocle3')
library(leidenbase)
library(monocle3)

```

```{r monocle3 pseudotemporal analysis}
## Please note that toggling between versions of R was necessary, as was saving workspace/global environment variables from the analyses that were initially ran above which used Seurat v.2.3.4
library(monocle3)
library(dplyr)
devtools::install_github("cole-trapnell-lab/monocle3", ref = "develop", force = TRUE)
#View(nichols_dr_seurat)
#To revert to old Seurat v2.3.4
#source("https://z.umn.edu/archived-seurat")


#Creates the CellDataSet from the updated seurat v.3 object. Building the monocle object also requires the subsetted meta.data -frame and a gene_short_name column to the updated seurat object. When converting the object to the monocle3, celldataset object, the gene common names populating the gene_short_name column which will be required to run functions in the monocle3 package.
nichols_data <- UpdateSeuratObject(nichols_TSNE)
#nichols_data <- as(as.matrix(nichols_TSNE@data), 'sparseMatrix')

gene_annotation <- as.data.frame(rownames(nichols_data@reductions[["pca"]]@feature.loadings), row.names = rownames(nichols_data@reductions[["pca"]]@feature.loadings))
colnames(gene_annotation) <- "gene_short_name"

cell_metadata <- as.data.frame(nichols_data@assays[["RNA"]]@counts@Dimnames[[2]], row.names = nichols_data@assays[["RNA"]]@counts@Dimnames[[2]])
colnames(cell_metadata) <- "barcode"

New_matrix <- nichols_data@assays[["RNA"]]@counts
New_matrix <- New_matrix[rownames(nichols_data@reductions[["pca"]]@feature.loadings), ]
expression_matrix <- New_matrix

cds_from_seurat <- new_cell_data_set(expression_matrix,
                                     cell_metadata = cell_metadata,
                                     gene_metadata = gene_annotation)


#cdsfromseurat <- saveRDS(cds_from_seurat, file = "cdsfromseurat.rds")
gene_names <- rownames(cds_from_seurat@assays@data@listData$counts)
#saveRDS(gene_names, "/Users/anthonypulvino/NicholsLabwork/scAlxWork/mAlxMapp/monocle_gene_names.rds")



## Preprocessing method, dimensionality reduction methods, and final clustering reflect same methods used in previous analyses in Seurat (PCA and UMAP). This provides some inter-analysis standardization. number of dimensions was also kept to the standard 10 used in previous analysis.

Nichols_cds <- preprocess_cds(cds_from_seurat, method = "PCA", num_dim = 55)


Nichols_cds <- reduce_dimension(Nichols_cds, reduction_method = 'UMAP', preprocess_method = "PCA")

Nichols_cds <- cluster_cells(Nichols_cds, cluster_method = "leiden", reduction_method = "UMAP")

Nichols_cds <- learn_graph(Nichols_cds, use_partition = TRUE)

Nichols_cds <- order_cells(Nichols_cds)
plot_cells(Nichols_cds, color_cells_by = "pseudotime")

par()
pc <- plot_cells(Nichols_cds, genes = c("alx3", "prdm1a","alcama", "barx1", "hoxc1a"))
pc_bar <- subset(pc, pc[[1]]$gene_short_name=="barx1")
```

```{r monocle3 top markers and subset to generate gene lists and plots for publication}
# Important to note that, although cluster placement has changed in monocle3's reduction of dimensions, data points identifying cell-type specific clusters are preserved. Run and view the following function and see for yourself.
Nichols_cds_markers <- top_markers(Nichols_cds, group_cells_by = "cluster", reference_cells = 2000, cores = 8)

old.idents <- c("1", "2", "3", "4")
new.idents <- c("Frontonasal", "Arches 1,2", "Posterior", "Melanocytes")
Nichols_cds_markers$cell_group <- plyr::mapvalues(x = Nichols_cds_markers$cell_group, from = old.idents, to = new.idents)
View(Nichols_cds_markers)

FN_markers <- subset(Nichols_cds_markers, Nichols_cds_markers$cell_group=="Frontonasal")
FN_markers <- FN_markers[order(FN_markers$marker_test_p_value),]
write.csv(FN_markers, file = "FN_markers_monocle3.csv")

A12_markers <- subset(Nichols_cds_markers, Nichols_cds_markers$cell_group=="Arches 1,2")
A12_markers <- A12_markers[order(A12_markers$marker_test_p_value),]
write.csv(A12_markers, file = "A12_markers_monocle3.csv")

POST_markers <- subset(Nichols_cds_markers, Nichols_cds_markers$cell_group=="Posterior")
POST_markers <- POST_markers[order(POST_markers$marker_test_p_value),]
write.csv(POST_markers, file = "POST_markers_monocle3.csv")

MELANO_markers <- subset(Nichols_cds_markers, Nichols_cds_markers$cell_group=="Melanocytes")
MELANO_markers <- MELANO_markers[order(MELANO_markers$marker_test_p_value),]
write.csv(MELANO_markers, file = "MELANO_markers_monocle3.csv")
View(MELANO_markers)

plot_cells(Nichols_cds, genes = FN_markers[,1][1:15])
plot_cells(Nichols_cds, genes = A12_markers[,1][1:15])
plot_cells(Nichols_cds, genes = POST_markers[,1][1:15])
plot_cells(Nichols_cds, genes = MELANO_markers[,1][1:15])

Nichols_cds <- learn_graph(Nichols_cds, use_partition = TRUE)
plot_cells(Nichols_cds_p)

Nichols_cds <- order_cells(Nichols_cds)
plot_cells(Nichols_cds, color_cells_by = "pseudotime", show_trajectory_graph = TRUE)
plot_cells(Nichols_cds, genes = c("cyp26b1", "alx3"), show_trajectory_graph = FALSE)

```

```{r method for renaming numbered clusters with actual name of NCC types based on gene lists}
# learn graph to fit principal plot charting pseudotime, order_cells should be used with the initial plot cells in mind. For this data set, selecting just the first point on the frontonasal cluster (#1) or the first , earliest derived cell on all clusters yielded the same charted pseudotemporal series.
Nichols_cds <- order_cells(Nichols_cds)

# quick way to get the cell-type specific clusters properly labeled for the data set used herein
colData(Nichols_cds)$assigned_cell_type <- as.character(clusters(Nichols_cds))
colData(Nichols_cds)$assigned_cell_type = dplyr::recode(colData(Nichols_cds)$assigned_cell_type,
                                                "1"="Frontonasal",
                                                "2"="Arches 1,2",
                                                "3"="Posterior",
                                                "4"="Melanocytes")


# re-plotting of clustered cells with specifically labeled identities by cell type assignment and pseudotemporal heat mapping. Note the same general trajectory is outlined specifying the pseudotemporal development of 
plot_cells(Nichols_cds, group_cells_by = "cluster", color_cells_by = "assigned_cell_type", group_label_size = 0.1)
plot_cells(Nichols_cds, group_cells_by = "cluster", color_cells_by = "pseudotime")


## code for 3D pseudotemporal plots
Nichols_cds_3d <- reduce_dimension(Nichols_cds, max_components = 3)
Nichols_cds_3d <- cluster_cells(Nichols_cds_3d)
Nichols_cds_3d <- learn_graph(Nichols_cds_3d)
plot_cells_3d(Nichols_cds_3d, color_cells_by="assigned_cell_type")
Nichols_cds_3d <- order_cells(cds_3d)
cds_3d_plot_obj <- plot_cells_3d(Nichols_cds, color_cells_by="pseudotime")

head(pData(Nichols_cds))

#Number of cells per cluster
table(Nichols_cds_p@clusters$UMAP$clusters)




```



```{r  subset of the FN cluster for trajectory analysis}
Nichols_cds_subset <- choose_cells(Nichols_cds)

Nichols_cds_subset <- preprocess_cds(Nichols_cds_subset)

Nichols_cds_subset <- reduce_dimension(Nichols_cds_subset, reduction_method = 'UMAP', preprocess_method = "PCA")

Nichols_cds_subset <- cluster_cells(Nichols_cds_subset, cluster_method = "leiden", reduction_method = "UMAP")

saveRDS(Nichols_cds_subset, "/Users/anthonypulvino/NicholsLabwork/scAlxWork/AppProjDir/Nichols_cds_subset.rds")

Nichols_cds_subsetMarkers <- top_markers(Nichols_cds_subset, group_cells_by = "cluster", reference_cells = 2000, cores = 8)

Nichols_cds_subset <- learn_graph(Nichols_cds_subset, use_partition = FALSE)
plot_cells(Nichols_cds_subset)
Nichols_cds_subset <- order_cells(Nichols_cds_subset, reduction_method = "UMAP")
plot_cells(Nichols_cds_subset, color_cells_by = "pseudotime", show_trajectory_graph = TRUE, cell_size = 1)
plot_cells(Nichols_cds_subset, genes = c("prrx1a"), cell_size = 1, show_trajectory_graph = TRUE)
plot_cells(Nichols_cds_subset, genes = c("alx4a"), cell_size = 1, show_trajectory_graph = TRUE)
plot_cells(Nichols_cds_subset, genes = c("alx3"), show_trajectory_graph = TRUE, color_cells_by = "pseudotime")

```

```{r monocle3 frontonasal trajectory inference --ATP}

## generate the CDS that we'll need to subset to make sure our app has it for input
gene_annotation <- as.data.frame(rownames(nichols_data@reductions[["pca"]]@feature.loadings), row.names = rownames(nichols_data@reductions[["pca"]]@feature.loadings))
colnames(gene_annotation) <- "gene_short_name"

cell_metadata <- as.data.frame(nichols_data@assays[["RNA"]]@counts@Dimnames[[2]], row.names = nichols_data@assays[["RNA"]]@counts@Dimnames[[2]])
colnames(cell_metadata) <- "barcode"

New_matrix <- nichols_data@assays[["RNA"]]@counts
New_matrix <- New_matrix[rownames(nichols_data@reductions[["pca"]]@feature.loadings), ]
expression_matrix <- New_matrix

cds_from_seurat <- new_cell_data_set(expression_matrix,
                                     cell_metadata = cell_metadata,
                                     gene_metadata = gene_annotation)

gene_names <- rownames(cds_from_seurat@assays@data$counts)

nichols_data <- UpdateSeuratObject(nichols_TSNE)
nichols_data <- as(as.matrix(nichols_TSNE@data), 'sparseMatrix')


## Preprocessing method, dimensionality reduction methods, and final clustering reflect same methods used in previous analyses in Seurat (PCA and UMAP). This provides some inter-analysis standardization. number of dimensions was also kept to the standard 10 used in previous analysis.
Nichols_cds <- preprocess_cds(cds_from_seurat, method = "PCA", num_dim = 55)
Nichols_cds <- reduce_dimension(Nichols_cds, reduction_method = 'UMAP', preprocess_method = "PCA")
Nichols_cds <- cluster_cells(Nichols_cds, cluster_method = "leiden", reduction_method = "UMAP")
Nichols_cds <- learn_graph(Nichols_cds, use_partition = TRUE)
Nichols_cds_markers <-top_markers(Nichols_cds)


## generate the frontonasal-y subsetted data that RShiny will take as part of it's input (see above)
Nichols_cds_subset <- choose_cells(Nichols_cds)
Nichols_cds_subset <- preprocess_cds(Nichols_cds_subset)
Nichols_cds_subset <- reduce_dimension(Nichols_cds_subset, reduction_method = 'UMAP', preprocess_method = "PCA")
Nichols_cds_subset <- cluster_cells(Nichols_cds_subset, cluster_method = "leiden", reduction_method = "UMAP")
#Nichols_cds_subsetMarkers <- top_markers(Nichols_cds_subset, group_cells_by = "cluster", reference_cells = 2000, cores = 8)
Nichols_cds_subset <- learn_graph(Nichols_cds_subset, use_partition = FALSE)

plot_cells(Nichols_cds_subset, genes = c("alx3"), cell_size = 1, show_trajectory_graph = TRUE)

table(Nichols_cds@clusters$UMAP$clusters)
plot_cells(Nichols_cds)

```


```{r Combined Mut+WT data analysis for Jamie 10.09.2020 SEURAT}

## All data

nichols_dr_seurat <- readRDS("/Users/anthonypulvino/NicholsLabwork/scAlxWork/AllData_app/nichols_dr_seurat_2019.rds")
nichols_dr_seurat3 <- UpdateSeuratObject(nichols_dr_seurat)

nichols_PCA <- RunPCA(nichols_dr_seurat3, dims.use = 1:10, do.fast = TRUE, seed.use = 123)

nichols_neighbors <- FindNeighbors(object = nichols_PCA, reduction = "pca", dims = 1:10, nn.eps = 0.5, verbose = FALSE, force.recalc = TRUE)

nichols_clusters_AllData <- FindClusters(object = nichols_neighbors, resolution = 0.2, n.start = 10, verbose = FALSE)
```

```{r Mut data only analysis for Jamie 10.09.2020 SEURAT then MONOCLE}

Idents(object=nichols_clusters_AllData) <- "orig.ident"

MUT_markers <- FindMarkers(nichols_clusters_AllData, ident.1 = "MUT")




#rename idents
nichols_clusters_AllData <- RenameIdents(nichols_clusters_AllData, "0" = "Arches 1,2", "1" = "Frontonasal", "2" = "Posterior", "3" = "Melanocytes")

saveRDS(nichols_clusters_AllData, "/Users/anthonypulvino/NicholsLabwork/scAlxWork/AllData_app/nichols_clusters_AllData.rds")
saveRDS(nichols_clusters_Mut, "/Users/anthonypulvino/NicholsLabwork/scAlxWork/AllData_app/nichols_clusters_Mut.rds")



## Mut-only data SEURAT
nicholsMut <- subset(nichols_dr_seurat3, idents = "MUT")

nichols_PCA_mut <- RunPCA(nicholsMut, dims.use = 1:10, do.fast = TRUE, seed.use = 123)

nichols_MutNeighbors <- FindNeighbors(object = nichols_PCA_mut, reduction = "pca", dims = 1:10, nn.eps = 0.5, verbose = FALSE, force.recalc = TRUE)

nichols_clusters_Mut <- FindClusters(object = nichols_MutNeighbors, resolution = 0.2, n.start = 10, verbose = FALSE)

nichols_clusters_Mut <- RenameIdents(nichols_clusters_Mut, "0" = "Arches 1,2", "1" = "Frontonasal", "2" = "Posterior", "3" = "Melanocytes")
DimPlot(nichols_clusters_Mut)
table(nichols_clusters_Mut@active.ident)
## Mut-only data MONOCLE3

gene_annotation_Mut <- as.data.frame(rownames(nichols_clusters_Mut@reductions[["pca"]]@feature.loadings), row.names = rownames(nichols_clusters_Mut@reductions[["pca"]]@feature.loadings))
colnames(gene_annotation_Mut) <- "gene_short_name"

cell_metadata_Mut <- as.data.frame(nichols_clusters_Mut@assays[["RNA"]]@counts@Dimnames[[2]], row.names = nichols_clusters_Mut@assays[["RNA"]]@counts@Dimnames[[2]])
colnames(cell_metadata_Mut) <- "barcode"

New_matrix_Mut <- nichols_clusters_Mut@assays[["RNA"]]@counts
New_matrix_Mut <- New_matrix_Mut[rownames(nichols_clusters_Mut@reductions[["pca"]]@feature.loadings), ]
expression_matrix_Mut <- New_matrix_Mut

cds_from_seurat_Mut <- new_cell_data_set(expression_matrix_Mut,
                                     cell_metadata = cell_metadata_Mut,
                                     gene_metadata = gene_annotation_Mut)

Nichols_cds_Mut <- preprocess_cds(cds_from_seurat_Mut, method = "PCA", num_dim = 21)

Nichols_cds_Mut <- reduce_dimension(Nichols_cds_Mut, reduction_method = 'UMAP', preprocess_method = "PCA")

Nichols_cds_Mut <- cluster_cells(Nichols_cds_Mut, cluster_method = "leiden", reduction_method = "UMAP")

Nichols_cds_Mut <- learn_graph(Nichols_cds_Mut, use_partition = TRUE)

table(Nichols_cds_Mut@clusters$UMAP$clusters)
MutMarkers <- top_markers(Nichols_cds_Mut) 
plot_cells(Nichols_cds_Mut)
plot_cells(Nichols_cds_Mut, group_label_size = 4) + ggplot2::theme(plot.title = element_text(size = 25))

## Arches Mutant Subset
Nichols_cds_A12subset <- choose_cells(Nichols_cds_Mut)
Nichols_cds_A12subset <- preprocess_cds(Nichols_cds_A12subset)
Nichols_cds_A12subset <- reduce_dimension(Nichols_cds_A12subset, reduction_method = 'UMAP', preprocess_method = "PCA")
Nichols_cds_A12subset <- cluster_cells(Nichols_cds_A12subset, cluster_method = "leiden", reduction_method = "UMAP")
#Nichols_cds_subsetMarkers <- top_markers(Nichols_cds_subset, group_cells_by = "cluster", reference_cells = 2000, cores = 8)
Nichols_cds_12subset <- learn_graph(Nichols_cds_A12subset, use_partition = FALSE)
plot_cells(Nichols_cds_12subset, show_trajectory_graph = TRUE, genes = "dlx5a", cell_size = 1)

plot_cells(Nichols_cds_A12subset, show_trajectory_graph = TRUE)
plot_cells(Nichols_cds_A12subset, show_trajectory_graph = TRUE, genes = "dlx5a", cell_size = 1)

```


```{r MUT and WT ARCHES1,2 traj infer 10.14.2020 MONOCLE}



ArchesWT <- readRDS("/Users/anthonypulvino/MichiHelp/Arches12_s3obj.rds")



## CREATING WT ARCHES CDS
gene_annotation_ArchesWT <- as.data.frame(rownames(ArchesWT@reductions[["pca"]]@feature.loadings), row.names = rownames(ArchesWT@reductions[["pca"]]@feature.loadings))
colnames(gene_annotation_ArchesWT) <- "gene_short_name"

cell_metadata_ArchesWT <- as.data.frame(ArchesWT@assays[["RNA"]]@counts@Dimnames[[2]], row.names = ArchesWT@assays[["RNA"]]@counts@Dimnames[[2]])
colnames(cell_metadata_ArchesWT) <- "barcode"

New_matrix_ArchesWT <- ArchesWT@assays[["RNA"]]@counts
New_matrix_ArchesWT <- New_matrix_ArchesWT[rownames(ArchesWT@reductions[["pca"]]@feature.loadings), ]
expression_matrix_ArchesWT <- New_matrix_ArchesWT

cds_from_seurat_ArchesWT <- new_cell_data_set(expression_matrix_ArchesWT,
                                     cell_metadata = cell_metadata_ArchesWT,
                                     gene_metadata = gene_annotation_ArchesWT)


Nichols_cds_ArchesWT <- preprocess_cds(cds_from_seurat_ArchesWT, method = "PCA", num_dim = 55)
Nichols_cds_ArchesWT <- reduce_dimension(Nichols_cds_ArchesWT, reduction_method = 'UMAP', preprocess_method = "PCA")
Nichols_cds_ArchesWT <- cluster_cells(Nichols_cds_ArchesWT, cluster_method = "leiden", reduction_method = "UMAP")
Nichols_cds_ArchesWT <- learn_graph(Nichols_cds_ArchesWT, use_partition = TRUE)


plot_cells(Nichols_cds_ArchesWT, genes = c("dlx5a", "hand2"), cell_size = 1.2, graph_label_size = 3, group_label_size = 4) + ggplot2::theme(plot.title = element_text(size = 25))



## CREATING MUTANT ARCHES CDS

ArchesMUT <- subset(nichols_clusters_Mut, idents = "0")

gene_annotation_ArchesMUT <- as.data.frame(rownames(ArchesMUT@reductions[["pca"]]@feature.loadings), row.names = rownames(ArchesMUT@reductions[["pca"]]@feature.loadings))
colnames(gene_annotation_ArchesMUT) <- "gene_short_name"

cell_metadata_ArchesMUT <- as.data.frame(ArchesMUT@assays[["RNA"]]@counts@Dimnames[[2]], row.names = ArchesMUT@assays[["RNA"]]@counts@Dimnames[[2]])
colnames(cell_metadata_ArchesMUT) <- "barcode"

New_matrix_ArchesMUT <- ArchesMUT@assays[["RNA"]]@counts
New_matrix_ArchesMUT <- New_matrix_ArchesMUT[rownames(ArchesMut@reductions[["pca"]]@feature.loadings), ]
expression_matrix_ArchesMUT <- New_matrix_ArchesMUT

cds_from_seurat_ArchesMUT <- new_cell_data_set(expression_matrix_ArchesMUT,
                                     cell_metadata = cell_metadata_ArchesMUT,
                                     gene_metadata = gene_annotation_ArchesMUT)


Nichols_cds_ArchesMUT <- preprocess_cds(cds_from_seurat_ArchesMUT, method = "PCA", num_dim = 55)
Nichols_cds_ArchesMUT <- reduce_dimension(Nichols_cds_ArchesMUT, reduction_method = 'UMAP', preprocess_method = "PCA")
Nichols_cds_ArchesMUT <- cluster_cells(Nichols_cds_ArchesMUT, cluster_method = "leiden", reduction_method = "UMAP")
Nichols_cds_ArchesMUT <- learn_graph(Nichols_cds_ArchesMUT, use_partition = TRUE)


plot_cells(Nichols_cds_ArchesMUT, genes = "alx3", cell_size = 1.2, graph_label_size = 3, group_label_size = 4) + ggplot2::theme(plot.title = element_text(size = 25))


#Nichols_cds_markers_MUT <- top_markers(Nichols_cds_MUT)



```




```{r ALL of the below blocks are rough draft work for the monocle and seurat apps}

library(shiny)
#install.packages("shinydashboard")
library(shinydashboard)
library(data.table)
#runExample("01_hello")
#install.packages("shinyWidgets")
library(shinyWidgets)

nichols_dr_seurat <- readRDS("/Users/anthonypulvino/NicholsLabwork/scAlxWork/nichols_dr_seurat_2019.rds") #whatever the location of the file download

gene_name <- Nichols_cds_markers[,1]

ui <- fluidPage(
                titlePanel("Frontonasal Feature Map Generator (from monocle3 analysis)"),
                
                
                tags$p("Please enter your gene of interest with no capitals or spaces into the search bar below and hit 'return' or click 'Go'."),
                
                
                #textOutput(outputId = "description"),
                
                searchInput(inputId = "gene_name",
                            label = "Enter Gene Name:", btnSearch = "Go"),
                
                tags$p("Generated below  is a feature map of the gene you searched based on its log10 expression values per cell across only frontonasal cells subsetted from our monocle3 (trajectory analysis) object."),
                
                #actionBttn(inputId = "Go", label = "Go"),
                
                plotOutput("plot_cells"),
                
                tags$a(href = "https://www.nicholslab.org/", "Nichols Lab Website"),
              
                )

server <- function(input, output, session) {
  
    #output$description <- renderText({
  #  isolate({"Please enter your gene of interest with no capitals or spaces into the search bar below and hit 'return'.   Generated below the search bar is a feature map of that gene based on its log10 expression values per cell across only   frontonasal cells subsetted from our monocle3 (trajectory analysis) object."})
   # })
    
    output$plot_cells <- renderPlot({
      
      title <- "plot_cells"
      plot_cells(Nichols_cds_subset, genes = input$gene_name, cell_size = 1.5)
        
      })
    
}

shinyApp(ui, server)
```




```{r}

Nichols_cds_subset <- readRDS("/Users/anthonypulvino/NicholsLabwork/scAlxWork/mAlxMapp/NicholsFNsubset.rds")
gene_name <- readRDS("/Users/anthonypulvino/NicholsLabwork/scAlxWork/mAlxMapp/WholeGeneList.rds")
#install.packages("shinyjs")
library(shinyjs)
library(shiny)

ui <- fluidPage(
  
                
  
                
  selectizeInput(inputId = "gene_name", label = "Enter Gene Name", choices = gene_name),
                
  plotOutput("plot_cells"),

  
)

server <- function(input, output) {

  data <- reactive({
    validate(
      need(input$gene_name != Nichols_cds_subset, "Please select a data set")
    )
    get(input$gene_name)
  })

  output$plot_cells <- renderPlot({
    plot_cells(Nichols_cds_subset, genes = data, cell_size = 1.5, graph_label_size = 3, group_label_size = 4) + theme(plot.title = element_text(size = 25))

  })

}

shinyApp(ui, server)
```

```{r}
nichols_data_obj <- saveRDS(nichols_data, file = "/Users/anthonypulvino/NicholsLabwork/scAlxWork/nichols_data_obj.rds")


library(shiny)
#install.packages("shinydashboard")
library(shinydashboard)
library(data.table)
#runExample("01_hello")
#install.packages("shinyWidgets")
library(shinyWidgets)

nichols_dr_seurat <- readRDS("/Users/anthonypulvino/NicholsLabwork/scAlxWork/nichols_dr_seurat_2019.rds") #whatever the location of the file download

gene_name <- Nichols_cds_markers[,1]

ui <- fluidPage(
                titlePanel("Frontonasal Feature Map Generator (from monocle3 analysis)"),
                
                
                tags$p("Please enter your gene of interest with no capitals or spaces into the search bar below and hit 'return' or click 'Go'."),
                
                
                #textOutput(outputId = "description"),
                
                searchInput(inputId = "gene_name",
                            label = "Enter Gene Name:", btnSearch = "Go"),
                
                tags$p("Generated below  is a feature map of the gene you searched based on its log10 expression values per cell across only frontonasal cells subsetted from our monocle3 (trajectory analysis) object."),
                
                #actionBttn(inputId = "Go", label = "Go"),
                
                plotOutput("plot_cells"),
                
                tags$a(href = "https://www.nicholslab.org/", "Nichols Lab Website"),
              
                )

server <- function(input, output, session) {
  
    #output$description <- renderText({
  #  isolate({"Please enter your gene of interest with no capitals or spaces into the search bar below and hit 'return'.   Generated below the search bar is a feature map of that gene based on its log10 expression values per cell across only   frontonasal cells subsetted from our monocle3 (trajectory analysis) object."})
   # })
    
    output$plot_cells <- renderPlot({
      
      title <- "plot_cells"
      plot_cells(Nichols_cds_subset, genes = input$gene_name, cell_size = 1.5)
        
      })

}

shinyApp(ui, server)
```



